File: examples/showcase/gtm/custom-code-bundles/gtm_pipeline_typescript/src/runtime/policy.ts
import type { GeneratedCapabilityRuntimeMetadata } from "../generated/runtime-target.js";

export type PolicyDecision = {
  decision: "allow" | "deny" | "clarify" | "approval_required";
  detail?: string;
  resolution?: Record<string, unknown>;
};

export async function evaluatePolicy(_context: {
  capability: GeneratedCapabilityRuntimeMetadata;
  params: Record<string, unknown>;
  rootPrincipal?: string;
}): Promise<PolicyDecision> {
  return {
    decision: "allow",
    detail: "GTM TypeScript native bundle evaluates actor and approval behavior in its backend adapter.",
  };
}

File: examples/showcase/gtm/custom-code-bundles/gtm_pipeline_go_native/extensions/policy.go
package extensions

import "{{ANIP_GO_MODULE_PATH}}/generated"

type PolicyDecision struct {
	Decision   string
	Detail     string
	Resolution map[string]any
}

type PolicyContext struct {
	Capability    generated.GeneratedCapabilityRuntimeMetadata
	Params        map[string]any
	RootPrincipal string
}

func EvaluatePolicy(_ PolicyContext) PolicyDecision {
	return PolicyDecision{
		Decision: "allow",
		Detail:   "GTM Go native bundle evaluates actor and approval behavior in its backend adapter.",
	}
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}
